baseapp/baseapp.go

```diff
@@ -130,33 +152,52 @@ type BaseApp struct { // nolint: maligned
 	abciListeners []ABCIListener
 }
 
+func (opt AppOptionOrdered) Order() OptionOrder { return opt.order }
+
+func (opt AppOptionFunc) Apply(app *BaseApp) { opt(app) }
+func (opt AppOptionFunc) Order() OptionOrder { return OptionOrderDefault }
+
+// StoreOption implements AppOption, and can be passed to the app constructor.
+func (opt StoreOption) Apply(app *BaseApp) { app.storeOpts = append(app.storeOpts, opt) }
+func (opt StoreOption) Order() OptionOrder { return OptionOrderDefault }
+
 // NewBaseApp returns a reference to an initialized BaseApp. It accepts a
 // variadic number of option functions, which act on the BaseApp to set
 // configuration choices.
 //
 // NOTE: The db is used to store the version number for now.
 func NewBaseApp(
-	name string, logger log.Logger, db dbm.DB, options ...func(*BaseApp),
+	name string,
+	logger log.Logger,
+	db dbm.DBConnection,
+	options ...AppOption,
```

Comment:
I think we should be careful about breaking the baseapp constructor in this way

Reply:
I tried to find a better way, but this made the most sense to me. The change to the signature is pretty minor, and the AppOption interface could give us flexibility for future changes.